### ISSUE TITLE
[Proposal] Arc Plugins

### ISSUE THREAD

**Issue by austinkelleher** [arc-repos/arc-workflows#28]
# Why?

As per our discussion early this week, it would be nice if arc provided hooks that can be handled via plugins. This is particularly useful for common tasks that would otherwise have to be implemented using npm scripts.

# How?

I propose that we create a new `Arc` class. This `Arc` class could contain useful information that can be passed to plugins. For example, we may want a plugin to only be activated in certain configured scenarios. The `Arc` class could expose the parsed config and other properties.

```js
class Arc {
  constructor (config) {
    this._config = config
    // ... And other stuff
  },

  getConfig () {
    return this._config
  }

  // ... And other stuff
}
```

Arc plugins could expose the following properties (and probably more):

- `name` {String}: Used to log out the plugin name to the console when being installed
- `async install(arc: Arc)` {Function): Called once the plugin is being registered. This can return a promise or nothing.
- `async beforeCreate(arc: Arc)` {Function): Called once before the create process begins. This can return a promise or nothing.
- `async afterCreate(arc: Arc)` {Function): Called once after the create process ends. This can return a promise or nothing.
- `async beforeDeploy(arc: Arc)` {Function): Called once before the deploy process begins. This can return a promise or nothing.
- `async afterDeploy(arc: Arc)` {Function): Called once after the deploy process ends. This can return a promise or nothing.

**example-arc-plugin.js**
```js
exports.name = 'example-arc'

exports.beforeCreate = (arc) => Promise.resolve()
exports.afterCreate = (arc) => Promise.resolve()

exports.beforeDeploy = (arc) => Promise.resolve()
exports.afterDeploy = (arc) => Promise.resolve()
```

Plugins do not necessarily have to return a promise. They can also return nothing:

**example-arc-plugin.js**
```js
exports.name = 'example-arc'

exports.beforeCreate = (arc) => {
  console.log('Just wanted to log some stuff before creation...')
}
```

# Usage

Arc plugins could be npm installed and registered in a `.arc` file.

```
npm install arc-marko --save-dev
```

```
@html
get /

@plugins
arc-marko
```

# Additional Thoughts

This part is not necessary, but I wanted to throw it out there regardless. We may find a reason to emit certain events with data that we do not want to be asynchronously handled. The `Arc` class could extend `EventEmitter` and emit events:

**example-arc-plugin.js**
```js
exports.name = 'example-arc'

exports.install = (arc) => {
  arc.on('some-event', (data) => {
    // idk
  })
}
```

**Comment by brianleroux** [comment]
yea I like a lot of this thinking --- will respond with some more detailed thoughts 🔜 !!

quick thoughts: 

- no need for a `class`; could just pass a plain obj statebag down the chain (ideally not writable/etc)
- I really don't like Promise for a host of reasons but I'm happy to support them in this sort of thing but only if a callback is not supplied (making the Promise code path opt-in for ppl who want those semantics)
- `@plugins` is a great idea for handling order dep plugins

**Comment by brianleroux** [comment]
one other thot! (and a reason to use a `class` or something with logic) would be to signal when a plugin is done its part of the processing
eg:

```arc
@plugins
marko-build
marko-deploy
```
```javascript
module.exports = {
  name: 'marco-deploy',
  beforeCreate(arc) {
    arc.next()
  }
}
```

or somethin like that

**Comment by austinkelleher** [comment]
@brianleroux Thanks for the feedback.

 I'd definitely like to support promises for the following use case:

```js
module.exports = {
  async beforeCreate () {
    await myFn()
    await myOtherFn()
  }
}
```

Additionally, I don't really care if we use `class` or a standard JavaScript object, but ultimately, I don't want to lock ourselves into something that isn't going to easily allow expansion. We will probably want to add functions later. I definitely think we should pass the parsed config to each function on the object though.

I'm not sure I am a huge fan of `arc.next()`. I think that plugins should probably be registered in series. I don't necessarily believe that plugins should ever rely on other plugins, but this is definitely safer and keeps the API simpler. If you return a Promise, it just becomes part of the chain. We can, of course, avoid `arc.next()` by just chaining the promises.

**Comment by brianleroux** [comment]
Err, ok that use case is fine but has nothing to do with Promises besides the syntax. (Aesthetics.) You can perform many sync ops with many techniques. I recognize front end devs are stoked on this syntax but it buys us little and can cost quite a bit when it comes to debugging, introspection and performance. 

For background, pls check out: https://www.joyent.com/blog/post-mortem-debugging-and-promises … this isn't always a concern and some of these concerns are better in Node 8 but even still you can end up with swallowed errors that have no line numbers. This is a desirable characteristic in use interfaces but not programmer tools where we need stack traces to do our job.

I don't like the `next` function either but it signals the next function is ready to run. Otherwise we have to _force_ everyone into a promise chain to achieve order dependent plugins. The alternative is there is no guaranteed order of execution to completion in plugins which means you can only have one plugin per event if you want any degree of determinism. Plugin architectures always end up with >1 plugin listening to an event. If they do anything that mutates state you are unfortunately opting into guaranteeing some degree of deterministic behavior. (LMK if that doesn't make sense and I can dig into that thinking farther.)  

notes/summary

- Promise needs to be opt-in (not required)
- Order dependent plugins are undesirable but we need to plan for them
- `class`: still not convinced! one can attach methods to a plain object 
- passing parsed config: why? is it problematic to make plugins parse the config themselves? what drawbacks are there if we do/don't?

**Comment by mlrawlings** [comment]
> but has nothing to do with Promises besides the syntax

@brianleroux not totally true: https://mathiasbynens.be/notes/async-stack-traces

**Comment by brianleroux** [comment]
1. I am very aware of the state of this
2. I am *not* advocating _against_ `Promise` --- I'm advocating they are opt-in
3. Lambda only supports Node 6.10 so we cannot support async/await (yet!)

**Comment by austinkelleher** [comment]
@brianleroux 

- Sorry for the confusion. I just meant that with async functions you don't have to explicitly return a promise.

- I don't feel strongly about using a `class` vs object.

- Maybe passing the parsed config is not necessary. I don't have a specific use case as to why you may want the parsed config, but I figured that we already had that information readily available. There is no reason to force plugins to re-parse what is already parsed.

- As for the `arc.next()` function, why don't we just pass a callback and inspect the function length. If the function length is 2, we know that they are using a callback and we will wait for for the callback to be done.

So the following would be supported:

**Return nothing**
```js
module.exports = {
  beforeCreate (arc) {
    // Return nothing
  }
}
```

**Return promise**
```js
module.exports = {
  beforeCreate (arc) {
    return Promise.resolve()
  }
}
```

**Callback provided**
```js
module.exports = {
  beforeCreate (arc, callback) {
    callback()
  }
}
```

**Comment by brianleroux** [comment]
No confusion and it's totally cool. Also: thx for sticking with me on this. The API will be better for it. 

Pretty sure we will need a way to signal continuation but willing to experiment with whatever. We've probably thought through enough of this for a first pass. I don't think I want to expose an interface around `create` events (yet) but the real value/interesting events are around `deploy` anyhow. 

Now for the really hard part. Naming conventions! 

- arc plugins are namespaced npm modules by `arc-plugin-` (?)
- arc plugins need to register their execution order in `.arc` under `@plugins` (I like this)
- first pass events to support `arc-before-deploy` and `arc-after-deploy` (?)

**Comment by austinkelleher** [comment]
@brianleroux No problem. This is something I want, and I'd be happy to work on it.

- I'm not a huge fan of the Babel convention that if you follow their plugin naming convention, you can reference the package in your `.babelrc` in a shorthand fashion. (e.g. `babel-plugin-transform-es2015-template-literals` can be referenced as `transform-es2015-template-literals`). I personally think that it causes a little confusion to new comers. I'd rather encourage that all packages follow our naming convention `arc-plugin-*`, but we don't do anything fancy for allowing shorthands. At least for now?

- I'm fine with only exposing `arc-before-deploy` and `arc-after-deploy` for now.

- I had another thought...We will probably want to pass config to each arc plugin. In `marko-starter` plugins, we allow users to extend the base config provided in the `marko-starter` project and pass along the project context (similar to `Arc`) to each of the hooks. For example (pseudocode):

**arc-plugin-marko.js**
```js
module.exports = {
  install () {
    extendBaseConfig({
      foo: String,
      description: 'The best config property ever'
    })
  }
}
``` 

The above method is nice because it allows us to ensure type safety at runtime (using [fashion-model](https://github.com/fashion-js/fashion-model)). `fashion-model` is able to provide great error messages to the user if they provide invalid config automatically. Something like: `Invalid type provided for "foo", should be a string.` 

Alternatively, in Lasso, we expose a single function and pass the config to it. We don't care about types until we start running the code. We do validation at that point.

**arc-plugin-marko.js**
```js
module.exports = function (arc, config) {
  if (!config) throw new Error('Config is required for arc-plugin-marko!')
  if (typeof config.foo !== 'string') {
    throw new Error('Invalid type provided for "foo", should be a string.')
  }

  return {
    beforeCreate () {
      ...
    }
  }
}
```

Thoughts?

**Comment by brianleroux** [comment]
yeah there will **def** be a requirement for per plugin configuration. where that goes is a good question. if in `.arc` then plugin authors are restricted to those types of primitives. 

```
@plugins
arc-plugin-marco
  key stringvalue
  anotherKey an array of values
  realBools true
  actualNums 100
```

some considerations

- its a really limited grammar but type friendly (json types anyhow) 
- idk if it'd be any good for more complex plugins (which maybe is a good thing also?)

First cut we could opt out of per plugin config / see how far we can get without it. More on the `.arc` format and, honestly very naive hand parser, here https://github.com/arc-repos/arc-parser

**Comment by brianleroux** [comment]
Also this is fun because maybe a pattern is emerging but we have something SUPER similar to this internally to decide what components to server render on which routes. We can assume GET because we don't render text/html with other verbs.

```
@plugins
arc-marco-lasso
  / some-index-component
  /pages page-screen
  /pages/:page page-detail-screen
```
The above would parse to the following `config`:

```javascript
{
"arc-marco-lasso": {
  "/": "some-index-component",
  "/pages": "page-screen",
  "/pages/:page": "page-detail-screen"
}
```

**Comment by brianleroux** [comment]
More fun: I have an early prototype for the [new deploy functionality here](https://github.com/brianleroux/arc-prototype-deploy). auto installs `node_modules` and copies `./src/shared` into each lambdas `node_modules/@architect/shared` folder before deploy.

**Comment by austinkelleher** [comment]
@brianleroux Thoughts on allowing arc to support `.arc.js` or something similar. In the Lasso plugin world, sometimes we only want to register a plugin in a certain scenario. For example, you may only want to register a plugin if `NODE_ENV` is set to production. There may be plugins that you only want to register when deploying to certain environments with arc too. I like the `.arc` syntax, but a JavaScript equivalent would provide a lot more power if needed. ESLint and others allow JavaScript config files.

In my opinion, we should allow config in the first pass to avoid a potential breaking change later. 

Plugin with config:

**.arc.js**
```js
module.exports = {
  html: [
    'get /',
    'get /pages'
  ],
  plugins: [
    {
      plugin: 'arc-marko',
      config: {
        foo: 'bar'
      }
    }
  ]
}
```

Plugin without config:


**.arc.js**
```js
module.exports = {
  html: [
    'get /'
  ],
  plugins: [ 'arc-marko' ]
}
```

**arc-plugin-marko.js**

```js
module.exports = function (arc, config) {
  // ... do some config validation here
  if (!config) throw new Error('Config is required for arc-plugin-marko!')
  if (typeof config.foo !== 'string') {
    throw new Error('Invalid type provided for "foo", should be a string.')
  }

  return {
    beforeCreate () {
      ...
    }
  }
}
```

**Comment by brianleroux** [comment]
For configuration I am not super supportive of adding additional formats before we port to another cloud. Too many uncertainties exist in the core use cases. Also not worried about waiting on this for reasons of breaking changes. Can't break a thing that didn't previously exist! This can totally be additive and non breaking to arc apps. 

(At the expense of plugin authors having to continue using whatever schemes they maybe opted themselves into previously.)

**Comment by brianleroux** [comment]
Ok! Update. Here's a prototype for us to pick apart: https://github.com/brianleroux/arc-prototype-deploy

Made the signature `beforeDepoy(config:String): Promise` …I couldn't think of a clean way to do this without having some form of concurrency signal which means either `Promise`, `callback` or both. I think a plugin interface should have a strong contract and only support one way of doing things and that way might as well be `Promise`. (Despite my now likely tiresome misgivings!)

**Comment by brianleroux** [comment]
Small update, we’ve been dogfooding this deploy script and it’s solid. Only hiccup is Node 8 is required and it expects lambdas to have package-lock.json present.

**Comment by austinkelleher** [comment]
Awesome @brianleroux. Will definitely try this out soon. Thanks for you work on this.

**Comment by austinkelleher** [comment]
Fixed in 2.0.0.
